libcaf_core/caf/flow/op/sample.hpp

```diff
@@ -0,0 +1,311 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/cow_vector.hpp"
+#include "caf/flow/observable_decl.hpp"
+#include "caf/flow/observer.hpp"
+#include "caf/flow/op/cold.hpp"
+#include "caf/flow/op/state.hpp"
+#include "caf/flow/step/all.hpp"
+#include "caf/flow/subscription.hpp"
+#include "caf/unit.hpp"
+
+#include <vector>
+
+namespace caf::flow::op {
+
+struct sample_input_t {};
+
+struct sample_emit_t {};
+
+///
+template <class T>
+class sample_sub : public subscription::impl_base {
+public:
+  // -- member types -----------------------------------------------------------
+
+  using input_type = T;
+
+  using output_type = T;
+
+  using select_token_type = int64_t;
+
+  // -- constants --------------------------------------------------------------
+
+  static constexpr size_t val_id = 0;
+
+  static constexpr size_t ctrl_id = 1;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  sample_sub(coordinator* parent, observer<output_type> out)
+    : parent_(parent), out_(std::move(out)) {
+    // nop
+  }
+
+  // -- properties -------------------------------------------------------------
+  coordinator* parent() const noexcept override {
+    return parent_;
+  }
+
+  bool running() const noexcept {
+    return state_ == state::running;
+  }
+
+  const error& err() const noexcept {
+    return err_;
+  }
+
+  bool pending() const noexcept {
+    return buf_.has_value();
+  }
+
+  bool can_emit() const noexcept {
+    return buf_.has_value() || has_shut_down(state_);
```

Comment:
I am a little bit confused about it to and I forgot to discuss regarding this. `buffer` has a similar function which probably needs to be renamed or the return value should be inverted.

https://github.com/actor-framework/actor-framework/blob/2caef0f0023279ef19a269ec449b90f676bc55d8/libcaf_core/caf/flow/op/buffer.hpp#L92C1-L94C4

Reply:
The story is a bit different for `buffer`, because it has a bit more going on. But for `sample`, we should stick to the simple model of emitting only on a tick. That being said, maybe there's room to simplify the `buffer` as well at some later time.